### ISSUE TITLE
TS2365: Operator '==' cannot be applied to types....when mutating array

### ISSUE THREAD

**Issue by kipz** [Microsoft/TypeScript#13697]
**TypeScript Version:**  2.1.5

**Code**

```ts
function  getLengthAsStr(strs: string[]): string {
  if(strs.length != 1){
     return "not 1"
  }
  //strs.length == 1
  strs.push("morestring")
 //strs.length == 2
  if(strs.length == 2){
    return "bigger than 1"
  }
  return "something else"
}

```

**Expected behaviour:**
Compiles without error or warning.

**Actual behaviour:**
Unexpected error: TS2365: Operator '==' cannot be applied to types '1' and '2'

**Comment by gcnew** [comment]
See https://github.com/Microsoft/TypeScript/issues/9998#issuecomment-235963457.

Basically TypeScript does not know that `push` mutates the array.

**Comment by kipz** [comment]
Thanks @gcnew - so is this considered a compiler bug to be fixed? It's just I noticed you didn't mark it as such.

**Comment by gcnew** [comment]
It is not considered a bug. The current consensus is that it's beneficial and desirable behaviour in the majority of cases. For those pesky mutable situations it can be worked aroud via indirection.

The problem is a hard one, because every function call may potentially invalidate the narrowing. As such, the pesimistic behaviour is really annoying and problematic in many more valid situations. The linked thread suggests several approches for improving the status-quo via inlining and an elaborate constraints scheme.

**Comment by kipz** [comment]
OK - thanks @gcnew. My personal feeling is that the pessimistic approach that works is all cases is preferable - but maybe I'm a bit old fashioned! :)

**Comment by larryaubstore** [comment]
Is there a way to disable these errors ?

**Comment by larryaubstore** [comment]
IMO, this typescript compiler issue should be reopen.

**Comment by fwouts** [comment]
I also got quite confused and was getting ready to submit a bug report because of this behaviour (see [the specific issue I ran into](https://github.com/fwouts/ts-bugreport-wrongtype/blob/master/example.ts)).

After reading through the comments, I now understand what's going on and I agree that there is no perfect way to fix it, however the error message could maybe point to an explanation and/or suggest alternatives? The trick of using re-fetching the mistyped variable via a method call isn't obvious at first.
